### PR TITLE
Refactor stripe samples commands

### PR DESCRIPTION
**Pull request by vcheung-stripe**
### Reviewers
r? @stripe/developer-products 
cc @stripe/developer-products

 ### Summary

- Hoisted the spinners, user IO, and error handling up to the cobra command
  - This removes IO concerns from the samples package
- Added new fields to the `Samples` struct to make it easier to work with 
- Updated and added a couple tests in the samples package

Verified manually by running stripe samples list and creating a variety of stripe samples with and without --force-refresh, with and without a path.

Motivated by https://github.com/stripe/stripe-cli/pull/646, which will be rebased on this.

### CODE REVIEW

**Comment by ob-stripe** [comment]
I've removed myself from the team. Re-rolling.

r? @stripe/developer-products

File: pkg/cmd/samples/create.go
Comment (pepin-stripe) [inline pkg/cmd/samples/create.go:67]: maybe out of scope of this diff, but to me this is an implementation detail that should live inside the package, not called here.

File: pkg/cmd/samples/create.go
Comment (pepin-stripe) [inline pkg/cmd/samples/create.go:143]: I think this function still has a lot of implementation details that should be moved into the samples package.

Things such as checking if files/dir exists, if the sample is already up to date, etc.

File: pkg/cmd/samples/create.go
Comment (pepin-stripe) [inline pkg/cmd/samples/create.go:283]: all those are fine here indeed. all the Prompts, Options, etc. (user input) should live here. The Cobra cmd should only have the responsibility to take in user input, and then pass it to the samples package as arguments to a "create" func.

File: pkg/cmd/samples/list.go
Comment (pepin-stripe) [inline pkg/cmd/samples/list.go:38]: yes this one looks great, except previous comment about having to do `samples.Samples{...}` seems that it's an implementation detail to me (not 100% convinced tho). Especially since we only pass defaults to it...

It should probably just be `samples.GetSamples("list")` without having to init a sample

**Review by pepin-stripe — APPROVED** [review]
So... I dont' wanna block gRPC but I feel like this cleanup could be a bit more complete.

It seems like this PR is moving cobra-related logic out of the samples package, which is great and probably enough to unblock gRPC.

What I think would be great to see if also cleaning up package-related logic out of the cobra command (see inline comments for more details).

Basically image calling `Samples.Create` or whatever from another place than the cobra function - all logic that would need to be duplicated from the command is probably good candidates to be extracted into the package.

File: pkg/cmd/samples/create.go
Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:143]: Agree, a single function that creates the sample and returns (path, error) would be ideal. I think this will make grpc implementation simpler as well.

File: pkg/cmd/samples/create.go
Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:67]: Do you mean the construction of the Samples object? I can change it to follow this pattern using `New` : https://github.com/stripe/stripe-cli/blob/3fbf545adac182d1374bbe04e5c84a22f0a04851/pkg/logtailing/tailer.go#L94-L107

Comment (pepin-stripe) [reply pkg/cmd/samples/create.go:67]: No, more about the fact that you have to initialize a `sample` object, only with default, to then call functions that feel like they could be package function instead of instance functions.

Basically instead of doing
```
sample := samples.Sample{}
sample.CreateSample()
```

it seems like the interface would be cleaner with
```
samples.CreateSample()
```
directly.

Comment (pepin-stripe) [reply pkg/cmd/samples/create.go:67]: This is not blocking by the way, mostly just nit/personal perf.

File: pkg/cmd/samples/create.go
Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:143]: I took a look at this and I think it's in scope because the grpc method will need this logic.

However, a problem with moving out the file existence check, sample is up to date check, etc, is that the command wants to print messages to the terminal while those checks are running. If we move out that logic, we would have to think about where to put the spinner and print statements.

If we want to move the logic to the samples package but keep the spinner and print statements in the cobra command, we could maybe use a channel? The `samples.Create` method could write a status to the channel: Downloading, Copying, Configuring, etc, and the cobra command could read the status and write the spinner and print statements asynchronously. What do you think of this approach?

Comment (pepin-stripe) [reply pkg/cmd/samples/create.go:143]: Agreed, I think channels is the way to go here to stream the output.

Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:143]: Sounds good I will add this!

**Comment by vcheung-stripe** [comment]
I split sample creation into two methods: 

- `samples.GetSampleConfig` returns a sample's config
  - The cobra command calls this so it knows what options to prompt the user in the terminal
  - The grpc method can call this to return the available options to a grpc client
- `samples.Create` returns a struct over a channel: current state of the routine, error, destination path, and post install message
  - The cobra command reads from the channel to print spinners and status messages
  - The grpc method doesn't care about the state, but it needs the destination path and post install message to return to the client

I'm not sure how clear this pattern is, but it will make implementing the grpc method simpler because there will be less duplication. ptal @pepin-stripe

File: pkg/cmd/samples/create.go
Comment (pepin-stripe) [inline pkg/cmd/samples/create.go:94]: you might want a `default` case for safety

**Review by pepin-stripe — APPROVED** [review]
LGTM! Nice work :)

File: pkg/cmd/samples/create.go
Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:95]: Added a default case, ptal @pepin-stripe

Comment (pepin-stripe) [reply pkg/cmd/samples/create.go:95]: did you push it? I don't see it.

Comment (vcheung-stripe) [reply pkg/cmd/samples/create.go:95]: Yep here it is:
https://github.com/stripe/stripe-cli/pull/652/files#diff-54c4064d2b4ba0c610ea0511c0faa5109f448125b2b307a35ee5c53939c0310aR117-R119

Comment (pepin-stripe) [reply pkg/cmd/samples/create.go:95]: wtf it didnt appear before I clicked that link! LGTM 🚢

**Comment by vcheung-stripe** [comment]
r-codeowners?

**Pull request merged**
